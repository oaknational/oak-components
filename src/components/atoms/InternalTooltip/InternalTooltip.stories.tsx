import React from "react";
import { Meta, StoryObj } from "@storybook/react";

import { OakBox } from "../OakBox";

import { InternalTooltip } from "./InternalTooltip";

import { colorArgTypes } from "@/storybook-helpers/colorStyleHelpers";
import { spacingArgTypes } from "@/storybook-helpers/spacingStyleHelpers";
import { borderArgTypes } from "@/storybook-helpers/borderStyleHelpers";
import { typographyArgTypes } from "@/storybook-helpers/typographyStyleHelpers";

const meta: Meta<typeof InternalTooltip> = {
  component: InternalTooltip,
  tags: ["autodocs"],
  title: "components/atoms/InternalTooltip",
  argTypes: {
    children: {
      control: {
        type: "text",
      },
    },
    ...colorArgTypes,
    ...spacingArgTypes,
    ...borderArgTypes,
    ...typographyArgTypes,
  },
  parameters: {
    controls: {
      include: [
        "children",
        "tooltipPosition",
        ...Object.keys(colorArgTypes),
        ...Object.keys(spacingArgTypes),
        ...Object.keys(borderArgTypes),
        ...Object.keys(typographyArgTypes),
      ],
    },
  },
  decorators: [
    (Story) => (
      <OakBox $ma="space-between-xxl">
        <Story />
      </OakBox>
    ),
  ],
  args: {
    children: "Hello there",
    tooltipPosition: "bottom-left",
  },
};
export default meta;

type Story = StoryObj<typeof InternalTooltip>;

export const Default: Story = {
  render: (args) => <InternalTooltip {...args} />,
};

export const WithStyling: Story = {
  render: (args) => (
    <InternalTooltip
      {...args}
      $background="lemon50"
      $ph="inner-padding-xl"
      $pv="inner-padding-m"
      $color="text-primary"
      $borderRadius="border-radius-m"
      $font="heading-light-7"
      $mb="space-between-xs"
    />
  ),
};
