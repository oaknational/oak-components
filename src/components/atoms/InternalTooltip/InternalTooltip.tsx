import React, { ReactNode } from "react";
import styled, { css } from "styled-components";

import { OakFlex, OakFlexProps } from "../OakFlex";

import { OakCombinedColorToken } from "@/styles";
import { parseColor } from "@/styles/helpers/parseColor";
import { ColorStyleProps } from "@/styles/utils/colorStyle";
import { responsiveStyle } from "@/styles/utils/responsiveStyle";
import { parseSpacing } from "@/styles/helpers/parseSpacing";

export type InternalTooltipProps = OakFlexProps & {
  children?: ReactNode;
  tooltipPosition?: "bottom-left" | "bottom-right" | "top-left" | "top-right";
  id?: string;
};

const StyledFlex = styled(OakFlex)`
  width: max-content;
  pointer-events: none;
`;

type StyledSvgProps = {
  $fill: ColorStyleProps["$background"];
  $tooltipPosition?: InternalTooltipProps["tooltipPosition"];
};

const ARROW_SIZE = parseSpacing("all-spacing-4");

const StyledSvg = styled.svg<StyledSvgProps>`
  position: absolute;
  ${({ $tooltipPosition }) => {
    switch ($tooltipPosition) {
      case "bottom-right":
        return css`
          top: -${ARROW_SIZE};
          right: ${parseSpacing("all-spacing-0")};
          transform: scale(-1, -1);
        `;
      case "top-right":
        return css`
          bottom: -${ARROW_SIZE};
          right: ${parseSpacing("all-spacing-0")};
          transform: scaleX(-1);
        `;
      case "top-left":
        return css`
          bottom: -${ARROW_SIZE};
          left: ${parseSpacing("all-spacing-0")};
        `;
      default:
        return css`
          top: -${ARROW_SIZE};
          left: ${parseSpacing("all-spacing-0")};
          transform: scaleY(-1);
        `;
    }
  }}
  ${responsiveStyle<StyledSvgProps, OakCombinedColorToken>(
    "fill",
    (props) => props.$fill,
    parseColor,
  )}
`;

/**
 * A primitive tooltip to be used as a basis for more opinionated UI components.
 */
export const InternalTooltip = ({
  children,
  $background = "black",
  $color = "text-inverted",
  tooltipPosition = "bottom-left",
  id,
  ...props
}: InternalTooltipProps) => {
  return (
    <StyledFlex
      role="tooltip"
      id={id}
      data-rac
      {...props}
      $position="relative"
      $background={$background}
      $color={$color}
      $maxWidth={["all-spacing-20", "all-spacing-22"]}
    >
      {children}
      <StyledSvg
        width={ARROW_SIZE}
        height={ARROW_SIZE}
        $fill={$background}
        $tooltipPosition={tooltipPosition}
        data-testid="tooltip-arrow"
      >
        <path d="M0 0H16L8 8L0 16V0Z" />
      </StyledSvg>
    </StyledFlex>
  );
};
