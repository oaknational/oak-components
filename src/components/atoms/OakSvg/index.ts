export { OakSvg } from "./OakSvg";
export type { OakSvgProps, OakSvgNames } from "./OakSvg";
