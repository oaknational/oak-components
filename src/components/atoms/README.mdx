# Atoms

## Rules for components in this folder

- Can import other atoms
- Imports from molecules and organisms are forbidden!
- Generic
- Unstyled
- Avoid embedded logic and state
- No sub-components
  (eg. Box)
