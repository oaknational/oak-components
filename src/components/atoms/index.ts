export * from "./OakBox";
export * from "./OakFlex";
export * from "./OakSpan";
export * from "./OakForm";
export * from "./OakImage";
export * from "./OakTypography";
export * from "./OakHeading";
export * from "./OakP";
export * from "./OakUL";
export * from "./OakOL";
export * from "./OakLI";
export * from "./OakLabel";
export * from "./OakIcon";
export * from "./OakGrid";
export * from "./OakGridArea";
export * from "./OakThemeProvider";
export * from "./OakMaxWidth";
export * from "./OakCloudinaryImage";
export * from "./OakKbd";
export * from "./OakGlobalStyle";
export * from "./OakAnchorTarget";
export * from "./OakFieldSet";
