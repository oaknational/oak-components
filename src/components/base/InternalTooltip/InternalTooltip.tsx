import React, { ReactNode } from "react";
import styled from "styled-components";

import { OakFlex, OakFlexProps } from "../OakFlex";

import { OakCombinedColorToken } from "@/styles";
import { parseColor } from "@/styles/helpers/parseColor";
import { ColorStyleProps } from "@/styles/utils/colorStyle";
import { responsiveStyle } from "@/styles/utils/responsiveStyle";

export type InternalTooltipProps = OakFlexProps & {
  isOpen: boolean;
  tooltip: ReactNode;
  children?: ReactNode;
};

const StyledFlex = styled(OakFlex)`
  width: max-content;
  pointer-events: none;
`;

type StyledSvgProps = {
  $fill: ColorStyleProps["$background"];
};

const StyledSvg = styled.svg<StyledSvgProps>`
  position: absolute;
  bottom: -16px;
  left: 0;
  ${responsiveStyle<StyledSvgProps, OakCombinedColorToken>(
    "fill",
    (props) => props.$fill,
    parseColor,
  )}
`;

export const InternalTooltip = ({
  isOpen,
  children,
  tooltip,
  $background = "black",
  $font = "heading-light-7",
  $color = "text-inverted",
  $borderRadius,
  ...props
}: InternalTooltipProps) => {
  return (
    <OakFlex $position="relative" $width="fit-content" $height="fit-content">
      {isOpen && (
        <OakFlex
          role="tooltip"
          $position="absolute"
          $top="all-spacing-0"
          $left="all-spacing-0"
          $transform="translateY(calc(-100% - 16px))"
          $zIndex="modal-dialog"
          $flexDirection="column"
        >
          <StyledFlex
            {...props}
            $position="relative"
            $background={$background}
            $color={$color}
            $font={$font}
            $btr={$borderRadius}
            $bbrr={$borderRadius}
            $maxWidth={["all-spacing-20", "all-spacing-22"]}
          >
            {tooltip}
            <StyledSvg width="16" height="16" $fill={$background}>
              <path d="M0 0H16L8 8L0 16V0Z" />
            </StyledSvg>
          </StyledFlex>
        </OakFlex>
      )}
      {children}
    </OakFlex>
  );
};
