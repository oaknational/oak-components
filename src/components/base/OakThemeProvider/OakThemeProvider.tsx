import React from "react";
import { ThemeProvider } from "styled-components";

import { OakTheme } from "@/styles";

export type OakThemeProviderProps = {
  theme: OakTheme;
  children: React.ReactNode;
};

export const OakThemeProvider = ({
  theme,
  children,
}: OakThemeProviderProps) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};
