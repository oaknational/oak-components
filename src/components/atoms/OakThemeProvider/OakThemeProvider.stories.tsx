import React from "react";
import { Meta, StoryObj } from "@storybook/react";

import { OakThemeProvider } from "./OakThemeProvider";

import { OakFlex } from "@/components/atoms/OakFlex";
import { oakDefaultTheme } from "@/styles";

const meta: Meta<typeof OakThemeProvider> = {
  component: OakThemeProvider,
  tags: ["autodocs"],
  title: "components/atoms/OakThemeProvider",
  argTypes: {},
  parameters: {
    controls: {
      include: [],
      sort: "none",
    },
  },
};
export default meta;

type Story = StoryObj<typeof OakThemeProvider>;

export const DefaultOakThemeProvider: Story = () => (
  <OakThemeProvider theme={oakDefaultTheme}>
    <OakFlex
      $background="bg-decorative1-main"
      $color="text-inverted"
      $ba="border-solid-m"
      $pa="spacing-24"
      $borderRadius="border-radius-l"
      $borderColor="border-primary"
    >
      Get a head-start on your lesson planning using quality-checked resources
      you can download and adapt for free.
    </OakFlex>
  </OakThemeProvider>
);

DefaultOakThemeProvider.args = {};
