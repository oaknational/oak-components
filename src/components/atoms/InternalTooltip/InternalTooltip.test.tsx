import { create } from "react-test-renderer";
import React from "react";
import "@testing-library/jest-dom";

import { InternalTooltip } from "./InternalTooltip";

import { OakThemeProvider } from "@/components/atoms/OakThemeProvider";
import renderWithTheme from "@/test-helpers/renderWithTheme";
import { oakDefaultTheme } from "@/styles";

describe(InternalTooltip, () => {
  it("matches snapshot", () => {
    const tree = create(
      <OakThemeProvider theme={oakDefaultTheme}>
        <InternalTooltip id={"tooltip"}>Hello there</InternalTooltip>
      </OakThemeProvider>,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("renders children", () => {
    const { getByText } = renderWithTheme(
      <InternalTooltip id={"tooltip"}>Hello there</InternalTooltip>,
    );

    expect(getByText("Hello there")).toBeInTheDocument();
  });

  it('positions the arrow based on the "tooltipPosition" prop', () => {
    const { rerender, getByTestId } = renderWithTheme(
      <InternalTooltip id={"tooltip"}>Hello there</InternalTooltip>,
    );

    expect(getByTestId("tooltip-arrow")).toHaveStyle("top: -1rem; left: 0rem");

    rerender(
      <InternalTooltip tooltipPosition="bottom-right" id={"tooltip"}>
        Hello there
      </InternalTooltip>,
    );

    expect(getByTestId("tooltip-arrow")).toHaveStyle("top: -1rem; right: 0rem");

    rerender(
      <InternalTooltip tooltipPosition="top-right" id={"tooltip"}>
        Hello there
      </InternalTooltip>,
    );

    expect(getByTestId("tooltip-arrow")).toHaveStyle(
      "bottom: -1rem; right: 0rem",
    );

    rerender(
      <InternalTooltip tooltipPosition="top-left" id={"tooltip"}>
        Hello there
      </InternalTooltip>,
    );

    expect(getByTestId("tooltip-arrow")).toHaveStyle(
      "bottom: -1rem; left: 0rem",
    );
  });
});
