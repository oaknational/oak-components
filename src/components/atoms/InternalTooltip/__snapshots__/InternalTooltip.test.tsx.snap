// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InternalTooltip matches snapshot 1`] = `
.c0 {
  position: relative;
  max-width: 22.5rem;
  color: #ffffff;
  background: #222222;
  font-family: Lexend,sans-serif;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c2 {
  width: -webkit-max-content;
  width: -moz-max-content;
  width: max-content;
  pointer-events: none;
}

.c3 {
  position: absolute;
  top: -1rem;
  left: 0rem;
  -webkit-transform: scaleY(-1);
  -ms-transform: scaleY(-1);
  transform: scaleY(-1);
  fill: #222222;
}

@media (min-width:750px) {
  .c0 {
    max-width: 40rem;
  }
}

<div
  className="c0 c1 c2"
  data-rac={true}
  id="tooltip"
  role="tooltip"
>
  Hello there
  <svg
    className="c3"
    data-testid="tooltip-arrow"
    height="1rem"
    width="1rem"
  >
    <path
      d="M0 0H16L8 8L0 16V0Z"
    />
  </svg>
</div>
`;
